File: dgx/server-node/src/routers/pdf.js
'use strict';

const express = require('express');
const fs = require('fs');
const path = require('path');
const multer = require('multer');
const { v4: uuidv4 } = require('uuid');

const { requireAuth } = require('../auth');
const config = require('../config');
const { jobStore } = require('../services/jobStore');
const { runJob } = require('../services/pdfService');

const router = express.Router();

// ── helpers ───────────────────────────────────────────────────────────────────

function uploadsDir() {
  const dir = path.resolve(config.DATA_DIR, 'uploads');
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function safePath(base, filename) {
  const resolved = path.resolve(base, filename);
  const baseResolved = path.resolve(base);
  if (!resolved.startsWith(baseResolved + path.sep) && resolved !== baseResolved) {
    throw new Error(`Path traversal detected: ${filename}`);
  }
  return resolved;
}

// ── multer setup ──────────────────────────────────────────────────────────────

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: config.MAX_UPLOAD_MB * 1024 * 1024 },
  fileFilter: (_req, file, cb) => {
    const isPdf =
      file.mimetype === 'application/pdf' ||
      file.originalname.toLowerCase().endsWith('.pdf');
    isPdf ? cb(null, true) : cb(new Error('Only PDF files are allowed'));
  },
});

// ── routes ────────────────────────────────────────────────────────────────────

router.post('/pdf/upload', requireAuth, upload.single('file'), (req, res) => {
  if (!req.file) {
    return res.status(400).json({ detail: 'No file uploaded' });
  }

  const buf = req.file.buffer;
  if (buf.length < 4 || buf.toString('ascii', 0, 4) !== '%PDF') {
    return res.status(400).json({ detail: 'Uploaded file does not appear to be a PDF' });
  }

  const jobId = uuidv4();
  const dest = safePath(uploadsDir(), `${jobId}.pdf`);
  fs.writeFileSync(dest, buf);
  jobStore.create(jobId);

  res.json({ job_id: jobId });
});

router.post('/pdf/jobs', requireAuth, express.json(), (req, res) => {
  const { job_id: jobId, operation, params = {} } = req.body || {};
  if (!jobId || !operation) {
    return res.status(400).json({ detail: "'job_id' and 'operation' are required" });
  }

  const job = jobStore.get(jobId);
  if (!job) {
    return res.status(404).json({ detail: 'job_id not found; upload PDF first' });
  }
  if (job.status !== 'pending') {
    return res.status(409).json({ detail: `Job already in state: ${job.status}` });
  }

  // Run asynchronously without blocking the event loop
  setImmediate(() => runJob(jobId, operation, params));

  res.json({ job_id: jobId, status: 'running' });
});

router.get('/pdf/jobs/:jobId', requireAuth, (req, res) => {
  const job = jobStore.get(req.params.jobId);
  if (!job) return res.status(404).json({ detail: 'Job not found' });

  res.json({
    job_id: job.jobId,
    status: job.status,
    result_file: job.resultFile || null,
    error: job.error || null,
  });
});

router.get('/pdf/download/:jobId', requireAuth, (req, res) => {
  const job = jobStore.get(req.params.jobId);
  if (!job) return res.status(404).json({ detail: 'Job not found' });
  if (job.status !== 'done') {
    return res.status(400).json({ detail: `Job status is '${job.status}'; not ready for download` });
  }
  if (!job.resultFile || !fs.existsSync(job.resultFile)) {
    return res.status(404).json({ detail: 'Result file not found on disk' });
  }

  const ext = path.extname(job.resultFile).toLowerCase();
  const contentType = ext === '.pdf' ? 'application/pdf' : 'text/plain; charset=utf-8';
  res.setHeader('Content-Type', contentType);
  res.download(job.resultFile);
});

router.delete('/pdf/jobs/:jobId', requireAuth, (req, res) => {
  const jobId = req.params.jobId;
  const job = jobStore.get(jobId);
  if (!job) return res.status(404).json({ detail: 'Job not found' });

  // Clean up upload — use safePath to reject path traversal
  try {
    const uploadFile = safePath(uploadsDir(), `${jobId}.pdf`);
    if (fs.existsSync(uploadFile)) {
      try { fs.unlinkSync(uploadFile); } catch { /* ignore */ }
    }
  } catch { /* safePath rejected the path — nothing to delete */ }

  // Clean up result — resultFile was written by the service (trusted), but
  // verify it still lives inside outputsDir before deleting
  if (job.resultFile) {
    try {
      const outDir = path.resolve(config.DATA_DIR, 'outputs');
      const resolved = path.resolve(job.resultFile);
      if (resolved.startsWith(path.resolve(outDir) + path.sep) && fs.existsSync(resolved)) {
        try { fs.unlinkSync(resolved); } catch { /* ignore */ }
      }
    } catch { /* ignore */ }
  }

  jobStore.delete(jobId);
  res.json({ deleted: jobId });
});

module.exports = router;

File: dgx/server/routers/pdf.py
import uuid
from pathlib import Path
from typing import Any

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel

from ..auth import require_auth
from ..config import get_settings
from ..services.job_store import job_store
from ..services.pdf_service import executor, run_job, uploads_dir, outputs_dir, safe_path

router = APIRouter()


class JobCreate(BaseModel):
    job_id: str
    operation: str
    params: dict[str, Any] = {}


@router.post("/pdf/upload")
async def upload_pdf(
    file: UploadFile = File(...),
    _: None = Depends(require_auth),
):
    settings = get_settings()
    max_bytes = settings.MAX_UPLOAD_MB * 1024 * 1024

    content = await file.read()
    if len(content) > max_bytes:
        raise HTTPException(413, f"File exceeds {settings.MAX_UPLOAD_MB} MB limit")
    if content[:4] != b"%PDF":
        raise HTTPException(400, "Uploaded file does not appear to be a PDF")

    job_id = str(uuid.uuid4())
    dest = safe_path(uploads_dir(), f"{job_id}.pdf")
    dest.write_bytes(content)
    job_store.create(job_id)
    return {"job_id": job_id}


@router.post("/pdf/jobs")
async def create_job(body: JobCreate, _: None = Depends(require_auth)):
    job = job_store.get(body.job_id)
    if job is None:
        raise HTTPException(404, "job_id not found; upload PDF first")
    if job.status != "pending":
        raise HTTPException(409, f"Job already in state: {job.status}")

    executor.submit(run_job, body.job_id, body.operation, body.params)
    return {"job_id": body.job_id, "status": "running"}


@router.get("/pdf/jobs/{job_id}")
async def get_job(job_id: str, _: None = Depends(require_auth)):
    job = job_store.get(job_id)
    if job is None:
        raise HTTPException(404, "Job not found")
    return {
        "job_id": job.job_id,
        "status": job.status,
        "result_file": job.result_file,
        "error": job.error,
    }


@router.get("/pdf/download/{job_id}")
async def download_result(job_id: str, _: None = Depends(require_auth)):
    job = job_store.get(job_id)
    if job is None:
        raise HTTPException(404, "Job not found")
    if job.status != "done":
        raise HTTPException(400, f"Job status is '{job.status}'; not ready for download")
    if not job.result_file:
        raise HTTPException(500, "Result file path missing")

    result_path = Path(job.result_file)
    if not result_path.exists():
        raise HTTPException(404, "Result file not found on disk")

    media_type = "application/pdf" if result_path.suffix == ".pdf" else "text/plain"
    return FileResponse(str(result_path), media_type=media_type, filename=result_path.name)


@router.delete("/pdf/jobs/{job_id}")
async def delete_job(job_id: str, _: None = Depends(require_auth)):
    job = job_store.get(job_id)
    if job is None:
        raise HTTPException(404, "Job not found")

    upload_file = uploads_dir() / f"{job_id}.pdf"
    if upload_file.exists():
        upload_file.unlink()

    if job.result_file:
        result_path = Path(job.result_file)
        if result_path.exists():
            result_path.unlink()

    job_store.delete(job_id)
    return {"deleted": job_id}
